### PR TITLE
[81_3] Enable switching between or closing documents through the tab pages

### PR DESCRIPTION
**Pull request by Minzihao**
## What
image

## How to test your changes?
Build and run, you can display it by the main menu "view" -> "Tab bar".

### CODE REVIEW

File: src/Plugins/Qt/qt_widget.cpp [hunk 1]
```diff
@@ -633,6 +633,13 @@ tree_view_widget (command cmd, tree data, tree actions) {
                                             cmd, data, actions);
   return abstract (wid);
 }
+widget
+tab_page_widget (url u, widget title, widget button, bool active) {
+  if (headless_mode) return headless_widget (); // can this work ?
```
Comment (Oyyko) [inline src/Plugins/Qt/qt_widget.cpp:638]: So can this work? 😂

**Comment by Oyyko** [comment]
Actually, you can create a git branch named "minzihao/81_3/Enable_switching" or something similar in **this** repo. I think this is good because
1. you don't need to fork the repo. And synchronize it with the original repo.
2. The reviewer can pull your code more easily.
@da-liii What do you think?

**Comment by Minzihao** [comment]
> Actually, you can create a git branch named "minzihao/81_3/Enable_switching" or something similar in **this** repo. I think this is good because
> 
> 1. you don't need to fork the repo. And synchronize it with the original repo. e.g Merge the original branch to your branch. You can just do `git checkout -b minzihao/81_10/something` with the newest mogan/branch-1.2, and then develop on the new branch minzihao/81_10/something.
> 2. The reviewer can pull your code more easily.
>    @da-liii What do you think?

Will do.

File: src/Plugins/Qt/qt_widget.cpp (hunk 1 above)
Comment (Minzihao) [reply src/Plugins/Qt/qt_widget.cpp:638]: Actually I don't know what is headless_mode🤔

Comment (da-liii) [reply src/Plugins/Qt/qt_widget.cpp:638]: headless mode is to use Mogan without launching the GUI, like use Mogan to do conversion: convert tmu format to HTML format

Comment (da-liii) [reply src/Plugins/Qt/qt_widget.cpp:638]: lolly (TeXmacs C++ library) -> moebius (TeXmacs primitives) -> typesetter -> editor (gui bundled)

If we dot not use Mogan to edit the tm/tmu file, we do not need the editor part. eg. LaTeX is headless by default.

**Comment by Oyyko** [comment]
> > Actually, you can create a git branch named "minzihao/81_3/Enable_switching" or something similar in **this** repo. I think this is good because
> > 
> > 1. you don't need to fork the repo. And synchronize it with the original repo. e.g Merge the original branch to your branch. You can just do `git checkout -b minzihao/81_10/something` with the newest mogan/branch-1.2, and then develop on the new branch minzihao/81_10/something.
> > 2. The reviewer can pull your code more easily.
> >    @da-liii What do you think?
> 
> Will do.

You can do that in the next PR.

File: src/Plugins/Qt/qt_widget.cpp (hunk 1 above)
Comment (Oyyko) [reply src/Plugins/Qt/qt_widget.cpp:638]: Based on the context, I think this function `tab_page_widget` will not be called during the headless mode, right?

Comment (da-liii) [reply src/Plugins/Qt/qt_widget.cpp:638]: > `tab_page_widget` will not be called during the headless mode, right?

It should be. Just remove it to see if the existing integration test still works fine. The unit tests and integration tests are using MoganResearch in a headless way.

I guess you are right. This line should be removed.

File: TeXmacs/progs/kernel/gui/menu-define.scm [hunk 2]
```diff
@@ -381,6 +381,10 @@
   (require-format x '(form-toggle :%2))
   `($form-toggle ,@(cdr x)))
 
+(define (gui-make-tab-page x)
+  (require-format x '(tab-page :%4))
```
Comment (da-liii) [inline TeXmacs/progs/kernel/gui/menu-define.scm:385]: 这个gui-make别的地方没啥问题，`:%4` 这个 4 是什么意思，这边加一下注释吧

File: TeXmacs/progs/kernel/gui/menu-widget.scm [hunk 3]
```diff
@@ -91,7 +91,8 @@
     (promise :%1)
     (ink :%1)
     (:menu-item-list)))
-  (:menu-item-list (:repeat :menu-item)))
+  (:menu-item-list (:repeat :menu-item))
+  (:tab-page (tab-page :%4)))
```
Comment (da-liii) [inline TeXmacs/progs/kernel/gui/menu-widget.scm:95]: 这里的`:%4`也许要加一下注释

File: TeXmacs/progs/kernel/gui/menu-widget.scm [hunk 4]
```diff
@@ -467,6 +468,21 @@
         (widget-balloon but twid)))
      (else but))))
 
+(define (make-tab-page p style bar?)
```
Comment (da-liii) [inline TeXmacs/progs/kernel/gui/menu-widget.scm:471]: 这里需要加一下注释，`p`,`style`,`bar?`都是什么意思

File: src/Plugins/Qt/QTMTabPage.cpp [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+void
+QTMTabPage::resizeEvent (QResizeEvent* e) {
+  int w= closeBtn->width ();
+  int h= closeBtn->height ();
+  int x= e->size ().width () - w - 8;
+  int y= e->size ().height () / 2 - h / 2;
+  closeBtn->setGeometry (x, y, w, h);
+}
```
Comment (da-liii) [inline src/Plugins/Qt/QTMTabPage.cpp:38]: 空行加一下

File: TeXmacs/progs/texmacs/texmacs/tm-server.scm [hunk 6]
```diff
@@ -179,6 +179,15 @@
              (when answ (buffer-close (current-buffer))))))
         (else (buffer-close (current-buffer)))))
 
+(tm-define (safely-kill-buffer-by-url tgt-buffer)
```
Comment (da-liii) [inline TeXmacs/progs/texmacs/texmacs/tm-server.scm:182]: 这里加一下注释，说明一下相关的代码逻辑，以及为什么需要这个safely kill

File: TeXmacs/progs/kernel/gui/menu-widget.scm (hunk 4 above)
Comment (da-liii) [reply TeXmacs/progs/kernel/gui/menu-widget.scm:471]: 这里的p这个参数命名太简洁了，建议改一个更好的参数名，这样注释可以写得更简单一些

不用管上边的代码风格，上面别的代码的代码风格是有问题的

File: src/Plugins/Qt/QTMTabPage.cpp [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+ * in the root directory or <http://www.gnu.org/licenses/gpl-3.0.html>.
+ ******************************************************************************/
+
+#include "QTMTabPage.hpp"
+
+/******************************************************************************
+ * QTMTabPage
+ ******************************************************************************/
+
+QTMTabPage::QTMTabPage (url u, QAction* title, QAction* _button, bool active)
```
Comment (da-liii) [inline src/Plugins/Qt/QTMTabPage.cpp:18]: 代码风格可以统一一下，建议参数加上 `p_` 前缀。我看了一下，墨干现在的Qt代码风格并不统一。使用`p_`前缀是常见的Qt代码风格，建议采用

Comment (da-liii) [inline src/Plugins/Qt/QTMTabPage.cpp:18]: 代码风格可以统一一下，建议参数加上 `p_` 前缀。我看了一下，墨干现在的Qt代码风格并不统一。使用`p_`前缀是常见的Qt代码风格，建议采用

File: src/Plugins/Qt/qt_widget.cpp (hunk 1 above)
Comment (Minzihao) [reply src/Plugins/Qt/qt_widget.cpp:638]: Ok I will remove this line.

File: TeXmacs/progs/kernel/gui/menu-define.scm (hunk 2 above)
Comment (Minzihao) [reply TeXmacs/progs/kernel/gui/menu-define.scm:385]: 已加注释，`:%4`对应前面的`require-format`意思是要求正好4个参数

File: TeXmacs/progs/kernel/gui/menu-widget.scm (hunk 3 above)
Comment (Minzihao) [reply TeXmacs/progs/kernel/gui/menu-widget.scm:95]: 已加注释

File: src/Plugins/Qt/QTMTabPage.cpp (hunk 7 above)
Comment (Minzihao) [reply src/Plugins/Qt/QTMTabPage.cpp:18]: 🆗

File: src/Plugins/Qt/QTMTabPage.cpp (hunk 5 above)
Comment (Minzihao) [reply src/Plugins/Qt/QTMTabPage.cpp:38]: 🆗

File: TeXmacs/progs/kernel/gui/menu-widget.scm (hunk 4 above)
Comment (Minzihao) [reply TeXmacs/progs/kernel/gui/menu-widget.scm:471]: Done✔.

**Comment by Minzihao** [comment]
All the suggestions are implemented @Oyyko @da-liii

**Comment by da-liii** [comment]
ci failed

**Comment by Minzihao** [comment]
81_3的3指的是这是我提交的第3次pr吗？ @da-liii

**Comment by da-liii** [comment]
> 81_3的3指的是这是我提交的第3次pr吗？

是的

**Comment by da-liii** [comment]
LGTM

**Comment by Minzihao** [comment]
@Oyyko

**Comment by Oyyko** [comment]
LGTM

**Review by Oyyko — APPROVED** [review]

**Comment by Oyyko** [comment]
@Minzihao You can decide to merge now or if you want to continue improving the code. This PR is approved and can be merged into branch-1.2 anytime.

**Pull request merged**